libstb/secvar/backend/edk2-compat-util.c

```diff
@@ -0,0 +1,601 @@
+		return OPAL_PERMISSION;
+
+	if (timestamp->minute > prev->minute)
+		return OPAL_SUCCESS;
+	if (timestamp->minute < prev->minute)
+		return OPAL_PERMISSION;
+
+	if (timestamp->second > prev->second)
+		return OPAL_SUCCESS;
+
+	/* Time less than or equal to is considered as replay*/
+	if (timestamp->second <= prev->second)
+		return OPAL_PERMISSION;
+
+	/* nanosecond, timezone, daylight and pad2 are meant to be zero */
+
+	return OPAL_SUCCESS;
+}
+
+int get_pkcs7(struct efi_variable_authentication_2 *auth, mbedtls_pkcs7 **pkcs7)
+{
+	char *checkpkcs7cert = NULL;
+	int len;
+	int rc;
+
+	len = get_pkcs7_len(auth);
+	if (len <= 0)
+		return OPAL_PARAMETER;		
+
+	if (!pkcs7)
+		return OPAL_PARAMETER;
+
+	*pkcs7 = malloc(sizeof(struct mbedtls_pkcs7));
+	if (!(*pkcs7))
+		return OPAL_NO_MEM;
+
+	mbedtls_pkcs7_init(*pkcs7);
+	rc = mbedtls_pkcs7_parse_der(
+			(const unsigned char *)auth->auth_info.cert_data,
+			(const unsigned int)len, *pkcs7);
+	if (rc) {
+		prlog(PR_ERR, "Parsing pkcs7 failed %04x\n", rc);
+		mbedtls_pkcs7_free(*pkcs7);
+		return rc;
+	}
+
+	checkpkcs7cert = zalloc(2048);
+	if (!checkpkcs7cert) {
+		mbedtls_pkcs7_free(*pkcs7);
+		return OPAL_NO_MEM;
+	}
+
+	rc = mbedtls_x509_crt_info(checkpkcs7cert, 2048, "CRT:", &((*pkcs7)->signed_data.certs));	
+	if (rc)
+		rc = OPAL_PARAMETER;
+	else
+		prlog(PR_DEBUG, "%s \n", checkpkcs7cert);
+
+	free(checkpkcs7cert);
+	mbedtls_pkcs7_free(*pkcs7);
+
+	return OPAL_SUCCESS;
+}
+
+int verify_signature(void *auth_buffer, char *newcert, uint64_t new_data_size,
+		     struct secvar *avar)
+{
+	struct efi_variable_authentication_2 *auth = NULL;
+	mbedtls_pkcs7 *pkcs7 = NULL;
+	mbedtls_x509_crt x509;
+	char *signing_cert = NULL;
+	char *x509_buf = NULL;
+	int signing_cert_size;
+	int rc;
+	char *errbuf;
+	int eslvarsize;
+	int eslsize;
+	int offset = 0;
+
+	if (!auth_buffer)
+		return OPAL_PARAMETER;
+
+	auth = auth_buffer;
+
+	/* Extract the pkcs7 from the auth structure */
+	rc  = get_pkcs7(auth, &pkcs7);
+	/* Failure to parse pkcs7 implies bad command */
+	if (rc < 0)
+		return OPAL_PARAMETER;		
+
+	prlog(PR_INFO, "Load the signing certificate from the keystore");
+
+	eslvarsize = avar->data_size;
+
+	/* Variable is not empty */
+	while (eslvarsize > 0) {
+		prlog(PR_INFO, "esl var size size is %d offset is %d\n", eslvarsize, offset);
+		if (eslvarsize < sizeof(EFI_SIGNATURE_LIST))
+			break;
+
+		/* Calculate the size of the ESL */
+		eslsize = get_esl_signature_list_size(avar->data + offset, eslvarsize);
+		/* If could not extract the size */
+		if (eslsize <= 0) {
+			rc = OPAL_PARAMETER;
+			break;
+		}
+
+		/* Extract the certificate from the ESL */
+		signing_cert_size = get_esl_cert(avar->data + offset, eslvarsize, &signing_cert);
+		if (signing_cert_size < 0) {
+			rc = signing_cert_size;
+			break;
+		}
+
+		mbedtls_x509_crt_init(&x509);
+		rc = mbedtls_x509_crt_parse(&x509, signing_cert, signing_cert_size);
+
+		/* If failure in parsing the certificate, try next */
```

Comment:
This is telling me that we can have an accepted variable that contains multiple ESLs, but some of them may not be valid data. Do we want to support this? That doesn't seem right to me

Reply:
if we are not pre-validating ESL certs format.. it is possible that someone cert file got corrupted and he created ESL from it.. but that may not mean all other certs are also corrupted

Reply:
So I think the answer is then that we should be checking the formats of the files when we verify the signature as well, even if its just a simple "parse_pkcs7" check on it. This can go in the `.validate()` function probably. 

With this change, then we can elevate this parse failure to be an (unexpected and unwanted) error

Reply:
Ideally we should.... it will not be simple check.. it will be getting the cert data from ESL and parsing that cert.. and do that for each ESL.. it will be like "validateESL()" function in itself.. to be called in ESL.. or before add new data..  I wonder if some of these should be part of the userspace util ?